### PR TITLE
Add tutorials using normalizing flows

### PR DESCRIPTION
**Pull request by francois-rozet**
This PR adds tutorials (SVI and VAE) using normalizing flows. The normalizing flow implementations are loaded from the [Zuko](https://github.com/probabilists/zuko) library. By the way, I think it could be interesting to add a "compatibility with external tools" section to the examples.

In addition (as discussed in #3297), this PR also updates the normalizing flow tutorial to mention that Pyro's built-in flow API is not developed anymore. I also fixed a mistake ($\det | \frac{\partial f}{\partial x} |$ should be $| \det \frac{\partial f}{\partial x} |$) in the math and cleaned up a bit.

### CODE REVIEW

**Comment by martinjankowiak** [comment]
@francois-rozet isn't `rsample_and_log_prob` zuko-specific?

**Comment by francois-rozet** [comment]
Indeed, but the class still works if `dist` does not implement `rsample_and_log_prob`. So it is not actually necessary to import `zuko`. Do you have a preference for `pyro.distributions` or `contrib.zuko` ?

BTW, `Transform.call_and_ladj` (also Zuko-specific) and `Distribution.rsample_and_log_prob` could be worth moving upstream (to torch itself), as they allow sampling and scoring without cache.

File: tutorial/source/normalizing_flows_intro.ipynb [hunk 1]
```diff
@@ -4,10 +4,12 @@
    "cell_type": "markdown",
    "metadata": {},
    "source": [
-    "# Normalizing Flows - Introduction (Part 1)\n",
+    "# Normalizing Flows - Introduction\n",
+    "\n",
+    "This tutorial introduces Pyro's built-in normalizing flows. It is independent of most of Pyro, but users may want to read about distribution shapes in the [Tensor Shapes Tutorial](http://pyro.ai/examples/tensor_shapes.html).\n",
+    "\n",
+    "> The development of Pyro's built-in flows has stopped in favor of external libraries. We recommend [Zuko](https://zuko.readthedocs.io) as it is compatible with Pyro, implements many flow architectures and is [well documented](https://zuko.readthedocs.io).\n",
```
[the remark this answers is not in the document]
Comment (francois-rozet) [reply tutorial/source/normalizing_flows_intro.ipynb:11]: I agree. I wanted to do that initially, but some libraries are unmaintained (`nflows`, `FlowTorch`) and some I don't know how they integrate with Pyro (`FrEIA`, `normflows`) as they don't follow PyTorch's distribution API. So should I just put all of them? I can also mention the [awesome-normalizing-flows](https://github.com/janosh/awesome-normalizing-flows) repo.

For the NF tuto, I only made minor edits, which either fix the error mentioned previously or simplify math notations (removing parentheses when clear from context). For example,

image


became

image

or

image


became

image

My editor also removed trailing white-spaces, but nothing else changed.

[the remark this answers is not in the document]
Comment (martinjankowiak) [reply tutorial/source/normalizing_flows_intro.ipynb:11]: i defer to your judgment as to which libraries are worth mentioning, though i'd err on the side of more. you can just make an explicit caveat like "although we note that some of these libraries may no longer be actively maintained and may have APIs that are not directly compatible with Pyro (and thus may require a custom wrapper for use in Pyro)."

**Comment by martinjankowiak** [comment]
@fritzo do you have a preference between `pyro.distributions` and `contrib.zuko`? i do not. the former is probably more discoverable...

**Comment by martinjankowiak** [comment]
note you can put a test for your contrib code in `tests/contrib`

File: pyro/contrib/zuko.py [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+# SPDX-License-Identifier: Apache-2.0
+
+import torch
+from torch import Size, Tensor
+
+import pyro
+
+
+class Zuko2Pyro(pyro.distributions.TorchDistribution):
+    r"""Wraps a Zuko (or PyTorch) distribution as a Pyro distribution."""
```
Comment (martinjankowiak) [inline pyro/contrib/zuko.py:11]: can you expand this docstring? link to zuko? etc

also the `(or PyTorch)` seems confusing since we have `pyro.distributions.TorchDistribution` for that

File: pyro/contrib/zuko.py [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+        elif self.has_rsample:
+            x = self.dist.rsample(shape)
+        else:
+            x = self.dist.sample(shape)
+
+        return x
+
+    def log_prob(self, x: Tensor) -> Tensor:
+        if x in self.cache:
+            return self.cache[x]
```
Comment (martinjankowiak) [inline pyro/contrib/zuko.py:41]: @fritzo can you please take a look at the caching logic?

File: tests/contrib/test_zuko.py [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+import torch
+
+import pyro
+from pyro.contrib.zuko import Zuko2Pyro
+from pyro.infer import SVI, Trace_ELBO
+from pyro.optim import Adam
+
+
+@pytest.mark.parametrize("multivariate", [True, False])
+def test_Zuko2Pyro(multivariate: bool):
```
Comment (martinjankowiak) [inline tests/contrib/test_zuko.py:15]: @fritzo anything else worth testing for?

File: tutorial/source/normalizing_flows_intro.ipynb [hunk 5]
```diff
@@ -4,10 +4,12 @@
    "cell_type": "markdown",
    "metadata": {},
    "source": [
-    "# Normalizing Flows - Introduction (Part 1)\n",
+    "# Normalizing Flows - Introduction\n",
+    "\n",
+    "This tutorial introduces Pyro's built-in normalizing flows. It is independent of most of Pyro, but users may want to read about distribution shapes in the [Tensor Shapes Tutorial](http://pyro.ai/examples/tensor_shapes.html).\n",
+    "\n",
+    "> The development of Pyro's built-in flows has stopped in favor of external libraries, such as [Zuko](https://github.com/probabilists/zuko), [nflows](https://github.com/bayesiains/nflows), [normflows](https://github.com/VincentStimper/normalizing-flows) or [FlowTorch](https://github.com/stefanwebb/flowtorch). Some of these libraries may have interfaces that are not directly compatible with Pyro. See the [SVI with flow guide](svi_flow_guide.ipynb) and [VAE with flow prior](vae_flow_prior.ipynb) tutorials for example usages of [Zuko](https://github.com/probabilists/zuko) within Pyro.\n",
```
Comment (martinjankowiak) [inline tutorial/source/normalizing_flows_intro.ipynb:11]: think this should point to flowtorch.ai?

File: tutorial/source/normalizing_flows_intro.ipynb [hunk 6]
```diff
@@ -4,10 +4,12 @@
    "cell_type": "markdown",
    "metadata": {},
    "source": [
-    "# Normalizing Flows - Introduction (Part 1)\n",
+    "# Normalizing Flows - Introduction\n",
+    "\n",
+    "This tutorial introduces Pyro's built-in normalizing flows. It is independent of most of Pyro, but users may want to read about distribution shapes in the [Tensor Shapes Tutorial](http://pyro.ai/examples/tensor_shapes.html).\n",
```
Comment (martinjankowiak) [inline tutorial/source/normalizing_flows_intro.ipynb:9]: can you add something like this to the top of both example notebooks?

"In this notebook we leverage normalizing flows implemented in Zuko, but similar functionality could also be obtained with another PyTorch-based normalizing library."

File: tutorial/source/svi_flow_guide.ipynb [hunk 7]
```diff
@@ -0,0 +1,236 @@
...
+    "        pyro.sample(\"z\", Zuko2Pyro(flow(x.flatten(-2))))"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## SVI\n",
+    "\n",
+    "We train our guide with a standard stochastic variational inference (SVI) pipeline. We use 16 particles to reduce the variance of the ELBO and clip the norm of the gradients to make training more stable."
```
Comment (martinjankowiak) [inline tutorial/source/svi_flow_guide.ipynb:113]: quite aggressive clipping?

File: tutorial/source/svi_flow_guide.ipynb [hunk 8]
```diff
@@ -0,0 +1,236 @@
...
+    "    ], dim=-1).unflatten(-1, (2, 2))\n",
+    "\n",
+    "    return pyro.distributions.MultivariateNormal(mu, cov)\n",
+    "\n",
+    "def model(x: Tensor):\n",
+    "    with pyro.plate(\"batch\", len(x)):\n",
+    "        z = pyro.sample(\"z\", prior)\n",
+    "\n",
+    "        with pyro.plate(\"obs\", 5):\n",
+    "            pyro.sample(\"x\", likelihood(z), obs=x.transpose(0, 1))"
```
Comment (martinjankowiak) [inline tutorial/source/svi_flow_guide.ipynb:62]: are all these transposes necessary?

**Comment by martinjankowiak** [comment]
@fritzo do we need to add zuko to `EXTRAS_REQUIRE`?

File: pyro/contrib/zuko.py (hunk 2 above)
Comment (francois-rozet) [reply pyro/contrib/zuko.py:11]: I don't think `TorchDistribution` does the same think. `Zuko2Pyro` takes any **instance** of distribution and makes it a valid Pyro distribution.

I'll add a page in the docs.

File: pyro/contrib/zuko.py (hunk 3 above)
Comment (francois-rozet) [reply pyro/contrib/zuko.py:41]: Note that I don't clear the cache as I consider that `Zuko2Pyro` objects should be temporary.

File: tutorial/source/svi_flow_guide.ipynb (hunk 7 above)
Comment (francois-rozet) [reply tutorial/source/svi_flow_guide.ipynb:113]: That's a value I use quite often, but I don't have much intuition about it.

File: tutorial/source/svi_flow_guide.ipynb (hunk 8 above)
Comment (francois-rozet) [reply tutorial/source/svi_flow_guide.ipynb:62]: Either there or in the guide.

**Comment by francois-rozet** [comment]
@martinjankowiak I have addressed your comments (at least the ones I could). I don't think it is necessary to add zuko in `EXTRAS_REQUIRE` as even `contrib/zuko.py` does not rely on it. However, it seems like `doctest` tries to run the example I added in the docstring and fails because it does not import `zuko`. I can modify the example to be a `code-block`, which should not be executed.

File: tutorial/source/svi_flow_guide.ipynb (hunk 8 above)
Comment (martinjankowiak) [reply tutorial/source/svi_flow_guide.ipynb:64]: it seems like this could be re-organized more elegantly to reduce the number of transposes

**Review by martinjankowiak — DISMISSED** [review]
lgtm. @fritzo could you please take a final pass, especially w.r.t. caching logic?

File: tutorial/source/svi_flow_guide.ipynb (hunk 8 above)
Comment (francois-rozet) [reply tutorial/source/svi_flow_guide.ipynb:64]: I think I can reduce the number of `transpose` from 2 to 1.

**Review by fritzo — COMMENTED** [review]
@francois-rozet thank you for cleaning up our existing tutorials and for adding this new tutorial! I think these will help lots of users provide exposure for Zuko. And thanks for your patience in this very slow holiday review cycle. I think we can push a Pyro 1.9 release soon after this merges.

LGTM, subject to @martinjankowiak's review and a couple nits.

File: pyro/contrib/zuko.py [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+`tutorial/vae_flow_prior.ipynb`.
+"""
+
+import torch
+from torch import Size, Tensor
+
+import pyro
+
+
+class Zuko2Pyro(pyro.distributions.TorchDistribution):
```
[the remark this answers is not in the document]
Comment (francois-rozet) [reply pyro/contrib/zuko.py:18]: Ahah I thought it was "cool" :laughing:, will fix

File: pyro/contrib/zuko.py (hunk 3 above)
Comment (francois-rozet) [reply pyro/contrib/zuko.py:72]: Note that the `cache` is only used when `x` is sampled with `rsample_and_log_prob`. But, yes I will mention it in the docstring.

File: tests/contrib/test_zuko.py (hunk 4 above)
Comment (francois-rozet) [reply tests/contrib/test_zuko.py:15]: I can add a "fake" `rsample_and_log_prob` method to the `dist` object and check that it runs with that.

```python
def rsample_and_log_prob(self, shape):
    x = self.rsample(shape)
    return x, self.log_prob(x)

dist.rsample_and_log_prob = rsample_and_log_prob
```

This would at least test that `ZukoToPyro` works as expected.

**Comment by francois-rozet** [comment]
I addressed your comments, and I think the tests have passed (went to bed before the end).

File: tests/contrib/test_zuko.py (hunk 4 above)
Comment (martinjankowiak) [reply tests/contrib/test_zuko.py]: sounds like a reasonable solution to me

**Comment by martinjankowiak** [comment]
lgtm apart from possibly adding a `rsample_and_log_prob` test

**Review by martinjankowiak — APPROVED** [review]

**Comment by francois-rozet** [comment]
> apart from possibly adding a rsample_and_log_prob test

I added it in the test.

File: tests/contrib/test_zuko.py (hunk 4 above)
Comment (francois-rozet) [reply tests/contrib/test_zuko.py]: It is done!

**Comment by martinjankowiak** [comment]
sorry i missed that thanks

**Comment by francois-rozet** [comment]
🎉

**Comment by martinjankowiak** [comment]
thanks @francois-rozet !!
